src/topkg_jbuilder.ml

```diff
@@ -1,11 +1,85 @@
 open StdLabels
 open Topkg
+open Fpath
+
+let uerror = Unix.error_message
+
+let rec mkdir dir =
+  let mode = 0o755 in
+  let aux d =
+    Log.debug (fun l -> l "mkdir %s" dir);
+    try Ok (Unix.mkdir d mode) with
+    | Unix.Unix_error (Unix.EEXIST, _, _) -> Ok ()
+    | Unix.Unix_error (e, _, _) ->
+      if d = dir
+      then R.error_msgf "create directory %s: %s" d (uerror e)
+      else R.error_msgf "create directory %s: %s: %s" dir d (uerror e)
+  in
+  OS.Dir.exists dir >>= function
+  | true  -> Ok ()
+  | false ->
+    let parent = Fpath.dirname dir in
+    OS.Dir.exists parent >>= function
+    | true  -> aux dir
+    | false -> mkdir parent >>= fun () -> aux dir
+
+let remove_file file =
+  Log.debug (fun l -> l "remove file %s" file);
+  try Ok (Unix.unlink file) with
+  | Unix.Unix_error (Unix.ENOENT, _, _) -> Ok ()
+  | Unix.Unix_error (e, _, _) ->
+    R.error_msgf "remove file %s: %s" file (uerror e)
+
+let remove_dir dir =
+  Log.debug (fun l -> l "remove dir %s" dir);
+  try Ok (Unix.rmdir dir) with
+  | Unix.Unix_error (Unix.ENOENT, _, _) -> Ok ()
+  | Unix.Unix_error (e, _, _) ->
+    R.error_msgf "remove directory %s: %s" dir (uerror e)
+
+let rmdir dir =
+  let rec aux = function
+    | [] -> Ok ()
+    | path :: todo ->
+      if not (Sys.file_exists path) then Ok ()
+      else if not (Sys.is_directory path) then
+        remove_file path >>= fun () ->
+        aux todo
+      else
+        OS.Dir.contents ~dotfiles:true ~rel:false path >>= fun paths ->
+        if paths = [] then
+          remove_dir path >>= fun () ->
+          aux todo
+        else
+          aux (paths @ path :: todo)
+  in
+  aux [dir]
+
+let rename src dst =
+  Log.debug (fun l -> l "rename %s into %s" src dst);
+  try Ok (Unix.rename src dst) with
+  | Unix.Unix_error (e, _, _) ->
+    R.error_msgf "rename %s into %s: %s" src dst (uerror e)
+
+let is_doc files =
+  (* [topkg doc] generate a call to build index.html *)
+  List.mem "doc/api.docdir/index.html" ~set:files
 
 let build =
   Pkg.build ()
-    ~cmd:(fun _c os _files ->
+    ~cmd:(fun _c os files ->
       let jbuilder = Conf.tool "jbuilder" os in
-      OS.Cmd.run @@ Cmd.(jbuilder % "build" % "--root" % "."))
+      Log.debug (fun l -> l "files=%s" (String.concat ", " files));
+      if is_doc files then (
+        OS.Cmd.run @@ Cmd.(jbuilder % "build" % "@doc" % "--root" % ".")
+        >>= fun () ->
+        let src = "_build" // "default" // "_doc" in
+        let dst = "_build" // "doc" // "api.docdir" in
+        rmdir dst >>= fun () ->
```

Comment:
OK. this is a lot of code for a workaround :(, but hopefully it's just temporary